bridgerelayer/chains/substrate/streamer.go

```diff
@@ -1,107 +1,177 @@
-// 				l.log.Error("Failed to fetch finalized header", "err", err)
-// 				retry--
-// 				time.Sleep(BlockRetryInterval)
-// 				continue
-// 			}
-
-// 			// Get block hash
-// 			finalizedHash, err := ss.Core.API.RPC.Chain.GetFinalizedHead()
-// 			if err != nil {
-// 				l.log.Error("Failed to fetch head hash", "err", err)
-// 				retry--
-// 				time.Sleep(BlockRetryInterval)
-// 				continue
-// 			}
-
-// 			// Sleep if the block we want comes after the most recently finalized block
-// 			if currentBlock > uint64(finalizedHeader.Number) {
-// 				ss.Core.Logger.Trace("Block not yet finalized", "target", currentBlock, "latest", finalizedHeader.Number)
-// 				time.Sleep(BlockRetryInterval)
-// 				continue
-// 			}
-
-// 			// Get hash for latest block, sleep and retry if not ready
-// 			hash, err := ss.Core.API.RPC.Chain.GetBlockHash(currentBlock)
-// 			if err != nil && err.Error() == ErrBlockNotReady.Error() {
-// 				time.Sleep(BlockRetryInterval)
-// 				continue
-// 			} else if err != nil {
-// 				ss.Core.Logger.Error("Failed to query latest block", "block", currentBlock, "err", err)
-// 				retry--
-// 				time.Sleep(BlockRetryInterval)
-// 				continue
-// 			}
-
-// 			// 1. Get transactions from block...
-
-// 			// 2. Send transaction and block to Router for packaging and relay...
-
-// 			currentBlock++
-// 			retry = BlockRetryLimit
-// 		}
-// 	}
-// }
+import (
+	"fmt"
+	"time"
+	"bytes"
+
+	log "github.com/sirupsen/logrus"
+	types "github.com/snowfork/go-substrate-rpc-client/types"
+	"github.com/snowfork/go-substrate-rpc-client/scale"
+	"github.com/snowfork/polkadot-ethereum/bridgerelayer/chains/ethereum"
+
+)
+
+// Streamer streams Substrate events
+type Streamer struct {
+	Core         *Core
+	EthRouter    *ethereum.Router
+	WebsocketURL string
+	BlockRetryLimit uint
+	BlockRetryInterval time.Duration
+}
+
+// NewStreamer returns a new substrate transaction streamer
+func NewStreamer(core *Core, er *ethereum.Router, websocketURL string, blockRetryLimit uint, blockRetryInterval uint) *Streamer {
+	return &Streamer{
+		Core:         core,
+		EthRouter:    er,
+		WebsocketURL: websocketURL,
+		BlockRetryLimit: blockRetryLimit,
+		BlockRetryInterval: time.Duration(blockRetryInterval) * time.Second,
+	}
+}
+
+// Start the streamer
+func (ss *Streamer) Start() error {
+	// Check whether latest is less than starting block
+	header, err := ss.Core.API.RPC.Chain.GetHeaderLatest()
+	if err != nil {
+		return err
+	}
+	if uint64(header.Number) < ss.Core.StartBlock {
+		return fmt.Errorf("starting block (%d) is greater than latest known block (%d)", ss.Core.StartBlock, header.Number)
+	}
+
+	err = ss.SubscribeBlocks(int(ss.BlockRetryLimit))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// SubscribeBlocks ...
+func (ss *Streamer) SubscribeBlocks(blockRetryLimit int) error {
+	var currentBlock = ss.Core.StartBlock
+	var retry = blockRetryLimit
+	for {
+		select {
+		default:
+			// No more retries, go to next block
```

Comment:
Is this comment accurate? If that happens, we are stopping the subscription, right?